### ISSUE TITLE
Gnosis external tests failing due stack too deep error when using via-ir and yul-optimizer 

### ISSUE THREAD

**Issue by r0qs** [ethereum/solidity#14082]
## Description

Compilation of Gnosis external tests started to fail due to stack too deep error when running with `ir-optimize-evm+yul` preset.

```
YulException: Cannot swap Variable var_operation with Variable _1: too deep in the stack by 4 slots in 
[ var_operation RET var_operation var_data_2122_length var_signatures_2137_mpos var_signatures_2137_mpos var_to
var_refundReceiver var_gasToken var_gasPrice var_baseGas var_safeTxGas var_value var_data_2122_offset 
var_refundReceiver var_to var_data_2122_length var_value var_gasToken var_gasPrice var_baseGas var_safeTxGas _1 ]

No memoryguard was present. Consider using memory-safe assembly only and annotating it via 'assembly ("memory-safe") { ... }'.

Error HH600: Compilation failed
```
The problem was introduced by this commit: https://github.com/safe-global/safe-contracts/commit/b9fdbde766eb0fee958cb9b12a3b3e5193fd96ae

## Environment

- Compiler version: `0.8.19`
- Compiler settings: `evmVersion: 'paris', viaIR: true,  optimizer: {enabled: true, details: {yul: true}}}`

## Steps to Reproduce

```
./test/externalTests/gnosis.sh native $(which solc)
```

**Comment by mmv08** [comment]
I'm from Safe, and I came to report this exact issue here and was surprised it is already being tracked. One interesting thing we found is that with solc 0.8.16 or 0.8.17, it works flawlessly.

**Comment by cameel** [comment]
Well, the "Stack too deep" error is easily affected by even small and seemingly unrelated changes in the piece of code that triggers it. It's not triggered by any single thing in high level code but by stack pressure from multiple small things accumulating and finally spilling over the 16 slot limit. Something we changed in code generation in 0.8.18 might have easily pushed it over if it was already close to that breaking point in your code. See https://github.com/ethereum/solidity/issues/13906#issuecomment-1417335621 for an example of such a situation. You could probably avoid it with trivial modifications around the change that finally triggered it, though often finding that spot is quite tedious and the compiler can't really point at it.

The high-level issue about "Stack too deep" is one of our roadmap tasks (#13721) and for unoptimized code #13972 will probably be a solution.

The issue here is only to patch it in our external tests. Preferably with an upstream patch. We were actually going to file an issue about that. @ekpyron suggested that just marking assembly blocks as `"memory safe"` on your side might do the trick.

**Comment by cameel** [comment]
By the way, @mmv08 there's actually a related topic we wanted to bring up to you (and other projects we use as external tests for the compiler).

Would you be willing to add a test run on your side with a bleeding edge compiler build to detect changes that could potentially affect your project? We're currently doing that on our side and the issue is that we often run into breakage that's not caused by the compiler itself but changes in dependencies, Hardhat, etc. Breakage is not that frequent from any single project but with 15 of them in our test suite we see it all the time. Since we're not in the JS ecosystem, it takes some effort to understand what happened and patch it. We thought that it would be more scalable if we could get upstream teams to regularly run their test suite run with the latest binary and report problems to us only when they're actually coming from the compiler. This way you get an early warning about changes that would break your code and we get only relevant reports to deal with. What do you think about it?

**Comment by mmv08** [comment]
@cameel, I will bring this to the team. It might make sense to do it because we run a GitHub action to detect compilation issues with 0.8.x (this is how we caught this issue)

**Comment by cameel** [comment]
Thanks! Let me know if they agree. We could then think how to set it up. We'd probably prepare some scripts from our side to make it easy to get the compiler and settings for the test.

**Comment by Killabits** [comment]
V 1.2

**Comment by cameel** [comment]
This was patched by #14086. Closing.
